client/signup/steps/website-content/index.tsx

```diff
@@ -171,6 +186,110 @@ function WebsiteContentStep( {
 	);
 }
 
+/**
+ * Hook that polls the /rest/v1.2/sites/<site fragment> API.
+ * It checks the site details for a valid DIFM purchase.
+ * A valid DIFM purchase has `.options.is_difm_lite_in_progress` set to `false`/`null`
+ * and `.options.difm_lite_site_options.pages` should be an array of strings.
+ *
+ * If the above conditions are met, the hook returns { isLoading: false, hasValidPurchase: true }.
+ * If the above conditions are not met, it retries the request MAXTRIES times,
+ * with a linear backoff. If the conditions are still not met, the hook returns
+ * { isLoading: false, hasValidPurchase: false }.
+ * The default return value is { isLoading: true, hasValidPurchase: false }
+ *
+ * @param {(SiteId | null)} siteId The current site ID.
+ * @returns {{
+ * 	isLoading: boolean;
+ * 	hasValidPurchase: boolean;
+ * }}
+ */
+function usePollSiteForDIFMDetails( siteId: SiteId | null ): {
```

Comment:
Wonder if we can move this entire code block outside of react. 

This would be a perfect use case to encapsulate logic inside a thunk action creator. After all this logic has nothing to do with view logic but it is related to the data layer almost exclusively.

WDYT?

Reply:
Also, this hook takes (heavy) inspiration from https://github.com/Automattic/wp-calypso/blob/trunk/packages/components/src/site-thumbnail/use-mshots-img.tsx#L34 🙂

Reply:
>Is there any benefit to adding actions like POLLING_START, POLLING_COMPLETE,..etc. and related intermediate states to the global redux store?

Nope we will not be maintaining  a lot of state. Just maybe a flag like `isLoggingInProgress`. And just two actions POLLING_STARTED, POLLING_COMPLETED. 

> This polling logic solves an extremely specific use case and is used for this component only. In the future, if this logic is required elsewhere, it can be easily moved to its own file.

This is I think a very good reason to move this logic inside a thunk with it's state living inside the websiteContent state slice. If in the future we think about a more generic use case then we can draw inspiration from this and create a proper generic hook for it.

> Having this as a hook co-located in the same file makes it easier to grok and understand the code without moving through multiple files like reducer.ts, action. ts, selector.ts,.

Actually all of this code almost completely will live inside in one function away from this component. So it will be 1 click at most to understand the logic. Because the rest of the redux (selectors, reducer) will be about updated 1 variable.

At other times, most of our logic will involve working with the website-content/index.tsx page. That means each time we want to fix a bug we will have to come to this page and be faced with this complex logic. The need to remove logic from this page comes from a place of reducing complexity as much as possible. And keep non like things away from this logic.

And beyond this purely from a performance standpoint we would be removing state from within react and simply moving it into a function.

---

No biggie here though. As you said this is more of a preference too in a sense. **Maybe we can move this into a separate file** 🤔

Reply:
> Also, this hook takes (heavy) inspiration from https://github.com/Automattic/wp-calypso/blob/trunk/packages/components/src/site-thumbnail/use-mshots-img.tsx#L34 🙂

Ah yes that makes perfect sense since there is a clear reusability win there. 💯 
If this hook was  actually the generic `pollForAPIState` which took in a URL and a success condition this hook would be perfect 😁.

Reply:
Sounds good to me! I have moved this into a separate file.